File: RandomNumbers/Random.cs
using UnityEngine;
using System.Collections;
using System.Security.Cryptography;

public class Random : MonoBehaviour
{
	uint N = 16384;
	uint ITERATIONS = 16384 * 1000;
	uint MIN;
	uint MAX;

	// Use this for initialization
	void Start ()
	{
		uint num = 100;
		float mean = 0f;

		/*ResetMinMax();
		for (int i = 0; i < num; ++i)
		{
			mean = mean + SystemRandom(); 
		}	
		mean = mean / (float) num;
		Debug.Log("Normal deviation mean: " + mean + ", iterations: " + num);
		Debug.Log("Min: " + MIN + ", Max: " + MAX);*/

		/*ResetMinMax();
		mean = 0f;
		for (int i = 0; i < num; ++i)
		{
			mean = mean + SystemCryptographyRandom(); 
		}	
		mean = mean / (float) num;
		Debug.Log("Normal deviation mean: " + mean + ", iterations: " + num);
		Debug.Log("Min: " + MIN + ", Max: " + MAX);*/

		/*ResetMinMax();
		mean = 0f;
		for (int i = 0; i < num; ++i)
		{
			mean = mean + UnityRandom(); 
		}	
		mean = mean / (float) num;
		Debug.Log("Normal deviation mean: " + mean + ", iterations: " + num);
		Debug.Log("Min: " + MIN + ", Max: " + MAX);*/

		ResetMinMax();
		mean = 0f;
		for (int i = 0; i < num; ++i)
		{
			mean = mean + WELL512Random(); 
		}	
		mean = mean / (float) num;
		Debug.Log("Normal deviation mean: " + mean + ", iterations: " + num);
		Debug.Log("Min: " + MIN + ", Max: " + MAX);
	}

	void ResetMinMax()
	{
		MIN = 2000;
		MAX = 0;
	}

	void GetMinMax(ushort[] array)
	{
		for (var i = 0; i < array.Length; ++i)
		{
			if (array[i] < MIN && array[i] != 0) MIN = array[i];	
			if (array[i] > MAX) MAX = array[i];
		}	
	}

	int CryptoServiceProviderSeed()
	{
		RNGCryptoServiceProvider random = new RNGCryptoServiceProvider ();

		byte[] seed = new byte[4];
		random.GetBytes (seed);

		return (int) (seed [0]) << 24 | (int) (seed [1]) << 16 | (int) (seed [1]) << 8 | seed [0];
	}


	float UnityRandom()
	{
		//UnityEngine.Random.InitState(Int) = Initializes the random number generator state with a seed.
		//UnityEngine.Random.Range(float, float) =

		int seed = CryptoServiceProviderSeed ();
		UnityEngine.Random.InitState(seed);

		ushort[] array = new ushort[N];
		for (int i = 0; i < ITERATIONS; ++i)
		{
			int index = Mathf.RoundToInt(UnityEngine.Random.Range(0f, (float) N-1));
			array[index] += 1;
		}

		GetMinMax(array);

		float mean = Mean(array, N);
		return Deviation(array, mean, N);
	}

	float WELL512Random()
	{
		int seed = CryptoServiceProviderSeed ();
		WELL512.init_well(seed);

		ushort[] array = new ushort[N];
		for (int i = 0; i < ITERATIONS; ++i)
		{
			int well = (int) WELL512.well512 ();

			int index = 0;
			if (well != int.MinValue) index = System.Math.Abs(well) % (int) N;
			array[index] += 1;
		}

		GetMinMax(array);

		float mean = Mean(array, N);
		return Deviation(array, mean, N);
	}

	float SystemRandom()
	{
		// Random()	= Initializes a new instance of the Random class, using a time-dependent default seed value.
		// Random(Int32) = Initializes a new instance of the Random class, using the specified seed value

		byte MASK = 63;
		ushort SHIFT = 8;

		byte[] randomArray = new byte[ITERATIONS*2];
		System.Random random = new System.Random();
		random.NextBytes (randomArray);

		ushort[] array = new ushort[N];
		for (int i = 0; i < randomArray.Length-1; i = i+2)
		{
			int rshort = (int) (randomArray[i] & MASK);
			int index = (rshort << SHIFT) | randomArray[i+1];
			array[index] += 1;
		}

		GetMinMax(array);

		float mean = Mean(array, N);
		return Deviation(array, mean, N);
	}

	float SystemCryptographyRandom()
	{
		//RandomNumberGenerator() = Initializes a new instance of RandomNumberGenerator.
		//GetBytes(Byte[]) = When overridden in a derived class, fills an array of bytes with a cryptographically
		// strong random sequence of values.

		byte MASK = 63;
		ushort SHIFT = 8;

		byte[] randomArray = new byte[ITERATIONS*2];
		RNGCryptoServiceProvider random = new RNGCryptoServiceProvider ();
		random.GetBytes (randomArray);

		ushort[] array = new ushort[N];
		for (int i = 0; i < randomArray.Length-1; i = i+2)
		{
			int rshort = (int) (randomArray[i] & MASK);
			int index = (rshort << SHIFT) | randomArray[i+1];
			array[index] += 1;
		}

		GetMinMax(array);

		float mean = Mean(array, N);
		return Deviation(array, mean, N);
	}

	float Deviation(ushort[] array, float mean, uint n)
	{
		float deviation = 0;
		for (int i = 0; i < n; i++)
		{
			deviation += Mathf.Pow(array[i] - mean, 2); 
		}
		deviation = Mathf.Sqrt(deviation / (float) (n - 1));
		return deviation;
	}

	float Mean(ushort[] array, uint n)
	{
		float mean = 0;
		for (int i = 0; i < n; i++)
		{
			mean += array[i]; 
		}	
		mean = mean / (float) n;
		return mean;
	}
}

File: RandomNumbers/random.js
var N = 16384;
var ITERATIONS = ITERATIONS = 16384 * 1000;

function GetRandom()
{	
	//console.log("N, ITERATIONS: " + N + ", " + ITERATIONS + ".");

	// Normal Randoms
	var array = Array.apply(null, Array(N)).map(function (x, i) { return 0; })
	
	for (var i = 0; i < ITERATIONS; i++)
	{ 
		var index = Math.floor((Math.random() * N) + 1);
		array[index] += 1;
	}

	var mean = Mean(array, N);
	var deviation = Deviation(array, mean, N);
	//console.log("Mean: " + mean + ", deviation: " + deviation);
	
	//WriteOutput(array, "test.txt");	
	return deviation;
}

function GetRandomCrypto()
{	
	//console.log("N, ITERATIONS: " + N + ", " + ITERATIONS + ".");

	var MASK = 63;
	var array = Array.apply(null, Array(N)).map(function (x, i) { return 0; })
	var cryptoBytes = crypto.randomBytes(ITERATIONS*2);

	for (var i = 0; i < cryptoBytes.length-1; i = i+2)
	{ 
		var index = ((cryptoBytes[i] & MASK) << 8) | cryptoBytes[i+1];
		array[index] += 1;
	}

	var mean = Mean(array, N);
	var deviation = Deviation(array, mean, N);
	//console.log("Mean: " + mean + ", deviation: " + deviation);

	//WriteOutput(array, "testCrypto.txt");
	return deviation;
}

function WriteOutput(data, filename)
{
	var fs = require("fs");
	var path = "C:\\" + filename;
	fs.writeFile(path, data, function(error) {
     		if (error) {
       			console.error("write error:  " + error.message);
     		} else {
      			console.log("Successful Write to " + path);
 		}
	});
}

function Deviation(array, mean, n)
{
	var deviation = 0;
	for (var i = 0; i < n; i++)
	{
		deviation += Math.pow(array[i] - mean, 2); 
	}
	deviation = Math.sqrt(deviation / (n - 1));
	return deviation;
}

function Mean(array, n)
{
	var mean = 0;
	for (var i = 0; i < n; i++)
	{
		mean += array[i]; 
	}	
	mean = mean / n;
	return mean;
}

function RandomAlgorithmsDeviationMean()
{
	var num = 100;
	var mean = 0;
	for (var i = 0; i < num; ++i)
	{
		mean = mean + GetRandom(); 
	}	
	mean = mean / num;
	console.log("Normal deviation mean: " + mean + ", iterations: " + num);

	mean = 0;
	for (i = 0; i < num; i++)
	{
		mean += GetRandomCrypto(); 
	}	
	mean = mean / num;
	console.log("Crypto deviation mean: " + mean + ", iterations: " + num);
}

RandomAlgorithmsDeviationMean();

